message_ix/macro.py

```diff
@@ -1,61 +1,471 @@
+        model_demand = self.s.var('DEMAND', filters={'level': 'useful'})
+        model_demand.rename(columns={'lvl': 'value', 'commodity': 'sector'},
+                            inplace=True)
+        model_demand = model_demand[idx + ['value']]
+        # get data provided in init year from data
+        demand_ref = self.data['demand_ref'].reset_index()
+        demand_ref['year'] = self.init_year
+        # combine into one value
+        demand = pd.concat([demand_ref, model_demand]).set_index(idx)['value']
+        if demand.isnull().any():
+            raise RuntimeError('NaN values found in demand calculation')
+        self.data['demand'] = demand
+        return demand
+
+    @lru_cache()
+    def _bconst(self):
+        price_ref = self.data['price_ref']
+        demand_ref = self.data['demand_ref']
+        rho = self._rho()
+        gdp0 = self._gdp0()
+        # TODO: automatically get the units here!!
+        bconst = price_ref / 1e3 * (gdp0 / demand_ref) ** (rho - 1)
+        self.data['bconst'] = bconst
+        return self.data['bconst']
+
+    @lru_cache()
+    def _aconst(self):
+        bconst = self._bconst()
+        demand_ref = self.data['demand_ref']
+        rho = self._rho()
+        gdp0 = self._gdp0()
+        k0 = self._k0()
+        kpvs = self.data['kpvs']
+        # TODO: why name this partmp??
+        partmp = (bconst * demand_ref ** rho).groupby(level='node').sum()
+        # TODO: automatically get the units here!!
+        aconst = ((gdp0 / 1e3) ** rho - partmp) / (k0 / 1e3) ** (rho * kpvs)
+        # want the series to only have index of node
+        self.data['aconst'] = aconst.reset_index(level='year', drop=True)
+        return self.data['aconst']
+
 
 def init(s):
-    for args in MACRO_INIT['sets']:
-        s.init_set(*args)
-    for args in MACRO_INIT['pars']:
-        s.init_par(*args)
-    # for args in MACRO_INIT['vars']:
-    #     print(args)
-    #     s.init_var(*args)
-    # for args in MACRO_INIT['equs']:
-    #     s.init_equ(*args)
+    for key, values in MACRO_INIT['sets'].items():
+        s.init_set(key, values)
+    for key, values in MACRO_INIT['pars'].items():
+        try:
+            s.init_par(key, values['idx'])
+        except:  # noqa: ignore=E722
+            pass  # already exists in the model, known for 'historical_gdp'
+    for key, values in MACRO_INIT['vars'].items():
+        if not s.has_var(key):
+            try:
+                # TODO: this seems required because for some reason DEMAND (and
+                # perhaps others) seem to already be listed in the java code,
+                # but still needs to be initialized in the python code.
+                # However, you cannot init it with dimensions, only with the
+                # variable name.
+                s.init_var(key, values)
+            except:  # noqa: ignore=E722
+                s.init_var(key)
+    for key, values in MACRO_INIT['equs'].items():
+        s.init_equ(key, values)
```

Comment:
@gidden, I needed to change this to work, it had some problems with existing equations in the scenario.
```
        try:
            s.init_equ(key, values)
        except:
            pass
```

Reply:
Hmm, yes, this is also related to https://github.com/iiasa/message_ix/pull/223#discussion_r312356449. I'll add it as a top-level item.